Support callbacks on multiple threads in DisplayConfigurationMonitor

mattwoodrow opened the pull request:
#### f517ea74ce7f0448baf00066743c606e1e0b4b50
<pre>
Support callbacks on multiple threads in DisplayConfigurationMonitor
<a href="https://bugs.webkit.org/show_bug.cgi?id=248993">https://bugs.webkit.org/show_bug.cgi?id=248993</a>

Reviewed by Cameron McCormack.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLANGLELinux.cpp:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp:
(WebCore::DisplayConfigurationMonitor::singleton):
(WebCore::DisplayConfigurationMonitor::addClient):
(WebCore::DisplayConfigurationMonitor::removeClient):
(WebCore::DisplayConfigurationMonitor::dispatchDisplayWasReconfigured):
* Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.h:
(WebCore::DisplayConfigurationMonitor::ClientEntry::operator== const):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::createGraphicsContextGL const):

Canonical link: <a href="https://commits.webkit.org/257616@main">https://commits.webkit.org/257616@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c566926390ded269a7aa3906267f5cb0e4b6fbe6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108852 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169082 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103479 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85971 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91954 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106768 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105244 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33947 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21861 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2513 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23376 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2439 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8593 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42854 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5237 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4300 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.h:
```diff
@@ -44,13 +46,23 @@ class DisplayConfigurationMonitor {
-    Vector<Client*> m_clients;
+    Lock m_lock;
+    struct ClientEntry {
+        Client* m_client;
```
Should this use the fancy new thread safe weak pointer thing?

mattwoodrow replied:
That's for things that are both refcounted and use weakptr. I'm not sure it helps us much here, since we're not refcounting Client objects.

heycam commented on Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.h:
```diff
@@ -44,13 +46,23 @@ class DisplayConfigurationMonitor {
+    Lock m_lock;
+    struct ClientEntry {
+        Client* m_client;
+        FunctionDispatcher* m_dispatcher { nullptr };
```
Please add a comment here to say that:
* the FunctionDispatcher lifetime must be at least as long as the Client's
* that the Client object must only be destroyed on the same thread / work queue as the one the FunctionDispatcher dispatches too, since we rely on that in the lambda we pass to dispatch() when we check the Client is still present in m_clients.

heycam commented on Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp:
```diff
@@ -42,29 +42,65 @@ DisplayConfigurationMonitor::~DisplayConfigurationMonitor() = default;
+        assertIsCurrent(*m_clients.at(index).m_dispatcher);
+    else
+        assertIsMainThread();
+    m_clients.removeFirst(ClientEntry { &client });
```
`m_clients.removeIndex(index);`, since we've already looked it up. (It will `ASSERT_WITH_SECURITY_IMPLICATIONS` with an invalid index.)

heycam commented on Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp:
```diff
@@ -42,29 +42,65 @@ DisplayConfigurationMonitor::~DisplayConfigurationMonitor() = default;
+                    {
+                        Locker locker { m_lock };
+                        if (!m_clients.contains(ClientEntry { client, dispatcher }))
+                            return;
```
Your `operator==` doesn't look at the dispatcher, so you don't need to pass that in.

Up to you, but I wonder if it would be nicer to use `containsIf`, `findIf`, etc., to avoid needing an `operator==` that doesn't check both fields.

I think we can get into a situation where just before we take this lock, a `Client` with a given address is removed and destroyed, and then a new one with the same address is created and added. I suppose it's not a huge deal: the `displayWasReconfigured` call on the new `Client` is unnecessary, but isn't going to harm anything. Probably worth commenting about though. (If you wanted to protect against this I guess `Client` would need to have some unique identifier on it.)

mattwoodrow committed:
Support callbacks on multiple threads in DisplayConfigurationMonitor
https://bugs.webkit.org/show_bug.cgi?id=248993

Reviewed by Cameron McCormack.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLANGLELinux.cpp:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp:
(WebCore::DisplayConfigurationMonitor::singleton):
(WebCore::DisplayConfigurationMonitor::addClient):
(WebCore::DisplayConfigurationMonitor::removeClient):
(WebCore::DisplayConfigurationMonitor::dispatchDisplayWasReconfigured):
* Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.h:
(WebCore::DisplayConfigurationMonitor::ClientEntry::operator== const):
* Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
(WebCore::createWebProcessGraphicsContextGL):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::createGraphicsContextGL const):

Canonical link: https://commits.webkit.org/257616@main
